### PR TITLE
Planet basemaps

### PR DESCRIPTION
**Pull request by taihak94**

### CODE REVIEW

File: Planet/PlanetBasemapIntegration.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+bbox = config['bbox']
+mongo_uri = config['db']
+
+# Setup db 
+client = pymongo.MongoClient(mongo_uri)
+db = client.louvre
+buildings_collection = db.buildings
+
+def get_tiles(ds, width=256, height=256):
+    nols, nrows = ds.meta['width'], ds.meta['height']
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: did you mean ncols?

File: Planet/PlanetBasemapIntegration.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        meta = inds.meta.copy()
+        for window, transform in get_tiles(inds):
+            meta['transform'] = transform
+            meta['width'], meta['height'] = window.width, window.height
+            outpath = os.path.join(folder,'{}_{}'.format(int(window.col_off), int(window.row_off)))
+            with rio.open(outpath, 'w', **meta) as outds:
+                outds.write(inds.read(window=window))
+
+
+def turn_to_json(filepath, mosaic_start_date, mosaic_end_date, url):
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: turn -> convert

turn what to json?

File: Planet/PlanetBasemapIntegration.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+                    
+                    # Transform shapes from the dataset's own coordinate
+                    # reference system to CRS84 (EPSG:4326).
+                    geom = rio.warp.transform_geom(
+                        dataset.crs, 'EPSG:4326', geom, precision=6)
+
+                    item['_id'] = 'planet_building_{}_{}_{}'.format(image_id, filename, counter)
+                    item['geometry'] = geom
+                    item['company'] = 'Planet'
+                    item['analyticsDeliveryTime'] = {
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: Update for new schema

File: Planet/PlanetBasemapIntegration.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+                        'end': datetime.utcfromtimestamp(int(round(mosaic_end_date.timestamp())))
+                    }
+                    item['analyticsInfo'] = {
+                        'url': url,
+                        'storage': 'Planet'
+                    }
+                    item['sourceImagesIds'] = [image_id]
+                    is_valid = analona.Building(item).validate()
+                    counter += 1
+                    if(is_valid == True):
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: if (is_valid)

File: Planet/PlanetBasemapIntegration.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+                        buildings_collection.update({ '_id': item['_id'] }, item, upsert = True)
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: replace_one

File: Planet/PlanetBasemapIntegration.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+                        print('pushed to db:', item['_id'])
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: should remove log?

File: Planet/PlanetBasemapIntegration.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+                    else:
+                        print('did not pass varification- {}'.format(is_valid))
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: verification

File: Planet/PlanetBasemapIntegration.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+    #Remove the folder of tiles
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: space after #

File: Planet/PlanetBasemapIntegration.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+        for window, transform in get_tiles(inds):
+            meta['transform'] = transform
+            meta['width'], meta['height'] = window.width, window.height
+            outpath = os.path.join(folder,'{}_{}'.format(int(window.col_off), int(window.row_off)))
+            with rio.open(outpath, 'w', **meta) as outds:
+                outds.write(inds.read(window=window))
+
+
+def turn_to_json(filepath, mosaic_start_date, mosaic_end_date, url):
+    folder = 'tmp_for_{}'.format(filepath.split('/')[1])
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: duplicate

File: Planet/PlanetBasemapIntegration.py [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+                    
+                shp = rio.features.shapes(
+                        mask, transform=dataset.transform)
+                # Extract feature shapes and values from the array.
+                for geom, _ in shp:
+                    item = {}
+                    
+                    # Transform shapes from the dataset's own coordinate
+                    # reference system to CRS84 (EPSG:4326).
+                    geom = rio.warp.transform_geom(
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: This is Polygon? MultiPolygon?

File: Planet/PlanetBasemapIntegration.py [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+            outpath = os.path.join(folder,'{}_{}'.format(int(window.col_off), int(window.row_off)))
+            with rio.open(outpath, 'w', **meta) as outds:
+                outds.write(inds.read(window=window))
+
+
+def turn_to_json(filepath, mosaic_start_date, mosaic_end_date, url):
+    folder = 'tmp_for_{}'.format(filepath.split('/')[1])
+    
+    #The first thing we do is split the original file
+    split_file(filepath)
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: tiles = split_tiles(filepath)

File: Planet/PlanetBasemapIntegration.py [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+
+
+def split_file(filepath):
+    folder = 'tmp_for_{}'.format(filepath.split('/')[1])
+    if os.path.exists(folder):
+        print("already split {}".format(filepath.split('/')[1]))
+        return
+    
+    os.mkdir(folder)
+    with rio.open(filepath) as inds:
```
[the remark this answers is not in the document]
Comment (taihak94) [reply Planet/PlanetBasemapIntegration.py]: in_ds where ds is dataStructure, but I'll change this

File: Planet/PlanetBasemapIntegration.py [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+
+def handle_quads_batch(session, mosaic_name, mosaic_start_date, mosaic_end_date, quad):
+    quad_items = quad['items']
+    for item in quad_items:        
+        download_quad(session, mosaic_name, mosaic_start_date, mosaic_end_date, item)
+
+
+def fetch_quads(session, mosaic_name, mosaic_start_date, mosaic_end_date, quad_items, download_directory):
+    if(not quad_items):
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: no need for () in simple if statements

File: Planet/PlanetBasemapIntegration.py [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+                if chunk:
+                    f.write(chunk)
+
+        # After we download the file locally, we can transform it and insert to louvre
+        raster_to_geojson(filepath, mosaic_start_date, mosaic_end_date, url)        
+    else:
+        print('no available download')
+
+
+def handle_quads_batch(session, mosaic_name, mosaic_start_date, mosaic_end_date, quad):
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: You made a mess of variable names
quad != quads != quad_items

Impossible to understand

File: Planet/PlanetBasemapIntegration.py [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+                        print("didn't pass verification- {}\nItem is {}".format(is_valid, item))
+    # Remove the folder of tiles
+    shutil.rmtree(folder)
+    # Remove the image because we don't need it anymore
+    os.remove(filepath)
+
+def download_quad(session, mosaic_name, mosaic_start_date, mosaic_end_date, quad):
+    directory = ''
+    url = quad['_links']['download']
+    if(url):
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: omit ()

File: Planet/PlanetBasemapIntegration.py [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+        for window, transform in get_tiles(img):
+            meta['transform'] = transform
+            meta['width'], meta['height'] = window.width, window.height
+            outpath = os.path.join(folder,'{}_{}'.format(int(window.col_off), int(window.row_off)))
+            with rio.open(outpath, 'w', **meta) as out_imgs:
+                out_imgs.write(img.read(window=window))
+
+
+def raster_to_geojson(filepath, mosaic_start_date, mosaic_end_date, url):
+    folder = 'tmp_for_{}'.format(filepath.split('/')[1])
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: is it always [1]?

File: Planet/PlanetBasemapIntegration.py [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+    
+    # The first thing we do is split the original file
+    split_file(filepath, folder)
+    
+    # Now the folder should contain all splited tiles.
+    for _, _, files in os.walk(folder):  
+        for filename in files:
+            file_route = os.path.join(folder, filename)
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: splitted_filepath

File: Planet/PlanetBasemapIntegration.py [hunk 18]
```diff
@@ -0,0 +1,177 @@
...
+                    geom = rio.warp.transform_geom(dataset.crs, 'EPSG:4326', geom, precision=6)
+
+                    item['_id'] = 'planet_building_{}_{}_{}'.format(image_id, filename, counter)
+                    item['geometry'] = geom
+                    item['company'] = 'Planet'
+                    item['observed_start'] = datetime.utcfromtimestamp(int(round(mosaic_start_date.timestamp())))
+                    item['observed_end'] = datetime.utcfromtimestamp(int(round(mosaic_end_date.timestamp())))
+                    item['analyticsInfo'] = {
+                        'url': url,
+                        'storage': 'Planet'
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: Put quad's download url here

File: Planet/PlanetBasemapIntegration.py [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+                shp = rio.features.shapes(mask, transform=dataset.transform)
+                # Extract feature shapes and values from the array.
+                for geom, _ in shp:
+                    item = {}
+                    
+                    # Transform shapes from the dataset's own coordinate
+                    # reference system to CRS84 (EPSG:4326).
+                    geom = rio.warp.transform_geom(dataset.crs, 'EPSG:4326', geom, precision=6)
+
+                    item['_id'] = 'planet_building_{}_{}_{}'.format(image_id, filename, counter)
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: collection name is plural 'buildings'

File: Planet/PlanetBasemapIntegration.py (hunk 16 above)
Comment (taihak94) [reply Planet/PlanetBasemapIntegration.py]: yep, I build the filepath in 137

File: Planet/PlanetBasemapIntegration.py (hunk 18 above)
Comment (taihak94) [reply Planet/PlanetBasemapIntegration.py]: that's what I put there...

File: Planet/PlanetBasemapIntegration.py [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+from urllib3.util import Retry
+
+from planet_downloader.basemap import basemap
+
+from itertools import product
+import rasterio as rio
+from rasterio import windows
+import rasterio.features
+import rasterio.warp
+
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: go over every import and make sure it is used

File: Planet/PlanetBasemapIntegration.py [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+                image_id = image_id.split('.')[0]
+                
+                # Init a counter for different IDs
+                counter = 0
+                # RGB Raster
+                if(dataset.count == 4):
+                    # Read the dataset's valid data mask as a ndarray.
+                    mask = dataset.read(3)
+                # BW Raster
+                elif(dataset.count == 2):
```
Comment (mzedayda) [inline Planet/PlanetBasemapIntegration.py]: no ()

**Pull request merged**